autospec/pkg_integrity.py

```diff
@@ -714,33 +714,68 @@ def attempt_key_import(keyid, key_fullpath):
     return False
 
 
-def parse_key(filename, pattern, verbose=True):
-    """Parse gpg --list-packet signature for pattern, return first match."""
-    args = ["gpg", "--list-packet", filename]
+def parse_gpg_packets(filename, verbose=True):
+    """Return a list with metadata about each packet from a GPG key or signature."""
+    args = ["gpg", "--list-packets", filename]
     try:
         out, err = Popen(args, stdout=PIPE, stderr=PIPE).communicate()
         if err.decode('utf-8') != '' and verbose is True:
             print(err.decode('utf-8'))
             return None
         out = out.decode('utf-8')
-        match = re.search(pattern, out)
-        return match.group(1).strip() if match else None
+        packets = []
+        for line in out.splitlines():
+            m = re.search(r'^# off=(\d+) ctb=.* tag=.* hlen=(\d+) plen=(\d+).*$', line)
+            if m:
+                packet = {}
+                packet["offset"] = int(m.group(1))
+                packet["length"] = int(m.group(2)) + int(m.group(3))
+                packets.append(packet)
+            m = re.search(r'^:signature packet:.* keyid ([0-9A-F]+)$', line)
+            if m:
+                packets[-1]["keyid"] = m.group(1)
+            m = re.search(r'^:user ID packet: "(.*) <(.+?)>"', line)
+            if m:
+                packets[-1]["user"] = m.group(1)
+                packets[-1]["email"] = m.group(2)
+        return packets
```

Comment:
I think I'd like this to be a little more resilient to unexpected data with hopefully some only minor changes.

Could we have a packet map defined outside the loop and then only append once the offset, length, keyid, user and email are found? Then if we find what looks to be a new packet before the old one is complete we just don't add that to the packets list. Thoughts?

Reply:
Not all packets will have all of those values set. But, yes, it would be more robust to initialize the packet list first, and then populate the packet info after we determine what data is present. I will make these changes.